src/test/util/setup_common.h

```diff
@@ -81,7 +81,7 @@ static constexpr CAmount CENT{1000000};
  * This just configures logging, data dir and chain parameters.
  */
 struct BasicTestingSetup {
-    node::NodeContext m_node;
+    node::NodeContext m_node; // keep as first member to be destructed last
```

Comment:
A foot-gun-proof way to do this would be to inherit from a class that holds a node::NodeContext.

Reply:
Something like this?

```diff
diff --git a/src/node/context.h b/src/node/context.h
index 31be308787..7ed1292180 100644
--- a/src/node/context.h
+++ b/src/node/context.h
@@ -29,6 +29,16 @@ class Init;
 class WalletLoader;
 } // namespace interfaces
 
+namespace {
+/**
+ * Dummy NodeContext to ensure that kernel::Context is always destructed last.
+ */
+struct BaseNodeContext {
+    //! libbitcoin_kernel context
+    std::unique_ptr<kernel::Context> kernel;
+};
+}
+
 namespace node {
 //! NodeContext struct containing references to chain state and connection
 //! state.
@@ -40,9 +50,7 @@ namespace node {
 //! and make code more modular and testable. The struct isn't intended to have
 //! any member functions. It should just be a collection of references that can
 //! be used without pulling in unwanted dependencies or functionality.
-struct NodeContext {
-    //! libbitcoin_kernel context
-    std::unique_ptr<kernel::Context> kernel;
+struct NodeContext : BaseNodeContext {
     //! Init interface for initializing current process and connecting to other processes.
     interfaces::Init* init{nullptr};
     std::unique_ptr<AddrMan> addrman;
diff --git a/src/test/util/setup_common.h b/src/test/util/setup_common.h
index 5c31cfc22b..87d5d4a24b 100644
--- a/src/test/util/setup_common.h
+++ b/src/test/util/setup_common.h
@@ -77,12 +77,20 @@ static inline bool InsecureRandBool() { return g_insecure_rand_ctx.randbool(); }
 
 static constexpr CAmount CENT{1000000};
 
+namespace {
+/**
+ * Dummy TestingSetup to ensure that node::NodeContext (and by extension
+ * kernel::Context) is always destructed last.
+ */
+struct BaseTestingSetup {
+    node::NodeContext m_node;
+};
+}
+
 /** Basic testing setup.
  * This just configures logging, data dir and chain parameters.
  */
-struct BasicTestingSetup {
-    node::NodeContext m_node; // keep as first member to be destructed last
-
+struct BasicTestingSetup : BaseTestingSetup {
     explicit BasicTestingSetup(const std::string& chainName = CBaseChainParams::MAIN, const std::vector<const char*>& extra_args = {});
     ~BasicTestingSetup();
 
```

Reply:
Yep, pretty much!

Sorry, I didn't mean to imply that needed to be done as part of this PR though. It might be worth taking a look at doing it generically with a template, which could help communicate the intent more clearly:
```c++
BasicTestingSetup : MustDestructLast<node::NodeContext>
```

But maybe that's over-engineering, idk.

Reply:
> A foot-gun-proof way to do this would be to inherit from a class that holds a node::NodeContext.

Would really encourage not doing this. There is no ambiguity about initialization order in the c++ language: earlier members are constructed first and destructed last, later members are constructed later and destructed earlier. Moving members into a base class to control what they are initialized here would seems mostly like a form of obfuscation.

It is true that if a future developer willingly wants to ignore a "this member should be first" comment it will be harder for them to do that if inheritance is involved, but it's true in general that if you obfuscate code it is harder to modify, and I don't think this logic leads to good places

Reply:
@ryanofsky Points taken. I think the intents/relationships need to be made clear somehow. Programmatically is one way, but maybe a comment works just as well without all the obfuscation.

In any case, this definitely shouldn't slow down this PR.

Reply:
I do agree relationship should be explicit if one thing depends on the other. Another safeguard that may be useful is if thing A needs to be destroyed after thing B, thing B destructor can assert that thing A is not destroyed yet (assuming relevant code will already have test coverage since it is part of the test setup)

Reply:
Okay, sounds like we're okay with leaving it as just a comment like in the latest push d87784ac87364fc977bbf9769c8bdb72dea8cbf9. Going to close this convo but feel free to re-open if further discussion is worth it.